File: app/src/main/java/ru/vincetti/vimoney/service/FileService.java
package ru.vincetti.vimoney.service;

import android.app.IntentService;
import android.content.Context;
import android.content.Intent;
import android.os.Environment;

import com.google.gson.Gson;

import java.io.File;
import java.io.FileOutputStream;

import io.reactivex.schedulers.Schedulers;
import ru.vincetti.vimoney.MyApp;
import ru.vincetti.vimoney.data.sqlite.AppDatabase;
import ru.vincetti.vimoney.settings.json.JsonFile;

public class FileService extends IntentService {
    public final static String SAVE_ACTION = "save-file-service";

    public FileService() {
        super("FileService");
    }

    @Override
    protected void onHandleIntent(Intent intent) {
        if (SAVE_ACTION.equals(intent.getAction())) {
            export();
        }
    }

    public void export() {
        Gson gson = new Gson();
        Context context = MyApp.getAppContext();

        AppDatabase.getInstance(context)
                .accountDao()
                .loadAllAccounts()
                .subscribeOn(Schedulers.io())
                .observeOn(Schedulers.io())
                .subscribe(
                        accountModels -> {
                            String accountJson = gson.toJson(accountModels);
                            writeFileExternalStorage(accountJson.getBytes(), JsonFile.FILE_NAME_ACCOUNTS);

                            AppDatabase
                                    .getInstance(context)
                                    .transactionDao()
                                    .loadAllTransactions()
                                    .subscribeOn(Schedulers.io())
                                    .observeOn(Schedulers.io())
                                    .subscribe(
                                            transactionModels -> {
                                                String transactionsJson = gson.toJson(transactionModels);
                                                writeFileExternalStorage(transactionsJson.getBytes(), JsonFile.FILE_NAME_TRANSACTIONS);

                                                MyApp.getAppContext().startService(new Intent(context, NotificationService.class)
                                                        .setAction(NotificationService.NOTIFICATION_SAVE_ACTION));
                                            });

                        }, throwable -> MyApp.getAppContext().startService(new Intent(context, NotificationService.class)
                                .setAction(NotificationService.NOTIFICATION_SAVE_ERROR_ACTION)));
    }

    public void writeFileExternalStorage(byte[] jsonBytes, String fileName) {
        //If it isn't mounted - we can't write into it.
        if (isExternalMounted()) {
            writeExternalFile(jsonBytes, fileName);
        }
    }

    private boolean isExternalMounted() {
        //Checking the availability state of the External Storage.
        return Environment.MEDIA_MOUNTED.equals(Environment.getExternalStorageState());
    }

    private void writeExternalFile(byte[] jsonBytes, String fileName) {
        //Create a new file that points to the root directory, with the given name:
        File jsonFile = new File(getExternalFilesDir(null), fileName);

        //the write operation
        FileOutputStream outputStream = null;
        try {
            outputStream = new FileOutputStream(jsonFile, false);
            outputStream.write(jsonBytes);
            outputStream.flush();
            outputStream.close();
        } catch (Exception e) {
            e.printStackTrace();
        }
    }
}

File: app/src/main/java/ru/vincetti/vimoney/service/FileService.kt
package ru.vincetti.vimoney.service

import android.app.IntentService
import android.content.Intent
import com.google.gson.Gson
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.launch
import ru.vincetti.vimoney.App
import ru.vincetti.vimoney.data.sqlite.AppDatabase
import ru.vincetti.vimoney.settings.json.JsonFile

class FileService : IntentService("FileService") {

    // использование своего job
    private val myJob = Job()
    private val myScope = CoroutineScope(Dispatchers.IO + myJob)

    companion object {
        const val SAVE_ACTION = "save-file-service"
    }

    override fun onHandleIntent(intent: Intent?) {
        if (SAVE_ACTION == intent?.action) export()
    }

    private fun export() {
        val gson = Gson()
        val context = App.getAppContext()

        val mDb = AppDatabase.getInstance(context!!)
        myScope.launch {
            val accounts = mDb.accountDao().loadAllAccounts()
            accounts?.let {
                val accountJson: String = gson.toJson(it)
                val fosAccount = App.context?.openFileOutput(JsonFile.FILE_NAME_ACCOUNTS, MODE_PRIVATE)
                fosAccount?.write(accountJson.toByteArray())
            }
            val transactions = mDb.transactionDao().loadAllTransactions()
            transactions?.let {
                val transactionsJson = gson.toJson(it)
                val fos = App.context?.openFileOutput(JsonFile.FILE_NAME_TRANSACTIONS, MODE_PRIVATE)
                fos?.write(transactionsJson.toByteArray())
            }
            App.getAppContext()?.startService(
                    Intent(App.context, NotificationService::class.java)
                            .setAction(NotificationService.NOTIFICATION_SAVE_ACTION))
        }
    }
}
